### PR TITLE
Use UnixFileStream's ReadAsync implementation on Windows when !IsAsync

### PR DESCRIPTION
**Pull request by stephentoub**
UnixFileStream's ReadAsync implementation uses a reusable IValueTaskSource implementation to avoid allocating a new work item on every read.  We can push that implementation down to OSFileStreamStrategy, and then use it for the Windows implementation of ReadAsync as well when IsAsync==false, rather than delegating to the base Stream implementation.

This PR almost entirely just moves code around.  The only change to logic is in RandomAccess.Windows.cs, to only set an offset into the NativeOverlapped if the SafeFileHandle is seekable; otherwise, it fails when used with pipes.

|    Method |         Toolchain |     Mean |    Error |   StdDev | Ratio |   Gen 0 |   Allocated |
|---------- |------------------ |---------:|---------:|---------:|------:|--------:|------------:|
|      Read | \main\corerun.exe | 36.19 ms | 0.568 ms | 0.531 ms |  1.00 |       - |        31 B |
|      Read |   \pr\corerun.exe | 35.71 ms | 0.296 ms | 0.263 ms |  0.99 |       - |        34 B |
|           |                   |          |          |          |       |         |             |
| ReadAsync | \main\corerun.exe | 45.84 ms | 0.360 ms | 0.337 ms |  1.00 | 90.9091 | 1,094,165 B |
| ReadAsync |   \pr\corerun.exe | 42.74 ms | 0.434 ms | 0.406 ms |  0.93 |       - |       256 B |

```C#
    private byte[] _buffer = new byte[4096];
    private FileStream _stream;

    [GlobalSetup]
    public void Setup()
    {
        byte[] bytes = new byte[40_000_000];
        new Random().NextBytes(bytes);

        string path = Path.GetTempFileName();
        File.WriteAllBytes(path, bytes);

        _stream = File.OpenRead(path);
    }

    [GlobalCleanup]
    public void Cleanup()
    {
        _stream.Close();
        File.Delete(_stream.Name);
    }

    [Benchmark]
    public void Read()
    {
        _stream.Position = 0;
        while (_stream.Read(_buffer) != 0) ;
    }

    [Benchmark]
    public async Task ReadAsync()
    {
        _stream.Position = 0;
        while (await _stream.ReadAsync(_buffer) != 0) ;
    }
```

cc: @adamsitnik

### CODE REVIEW

**Review by adamsitnik — APPROVED** [review]
LGTM, thank you @stephentoub !

**Comment by adamsitnik** [comment]
I am going to merge it now so I can get updated numbers for the blog post ;)

**Comment by adamsitnik** [comment]
I am not sure if this was intentional, but after moving all the logic from `Unix` to `OS` strategy it's now much easier to work with the code on Windows. Refactoring in VS renames all usages of methods, and it's less likely for the code to compile fine on Windows and fail on Unix.  👍

File: src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Windows.cs [hunk 1]
```diff
@@ -695,8 +695,11 @@ private static NativeOverlapped GetNativeOverlappedForSyncHandle(SafeFileHandle
             Debug.Assert(!handle.IsAsync);
 
             NativeOverlapped result = default;
-            result.OffsetLow = unchecked((int)fileOffset);
-            result.OffsetHigh = (int)(fileOffset >> 32);
+            if (handle.CanSeek)
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Windows.cs:698]: > this should not be needed

But it is.  Try removing it ;-)

File: src/libraries/System.Private.CoreLib/src/System/IO/Strategies/SyncWindowsFileStreamStrategy.cs [hunk 2]
```diff
@@ -20,45 +17,5 @@ internal SyncWindowsFileStreamStrategy(string path, FileMode mode, FileAccess ac
...
-        public override ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken = default)
-        {
-            // If we weren't opened for asynchronous I/O, we still call to the base implementation so that
-            // Write is invoked asynchronously.  But if we have a byte[], we can do so using the base Stream's
-            // internal helper that bypasses delegating to BeginWrite, since we already know this is FileStream
-            // rather than something derived from it and what our BeginWrite implementation is going to do.
-            return MemoryMarshal.TryGetArray(buffer, out ArraySegment<byte> segment) ?
-                new ValueTask(BeginWriteInternal(segment.Array!, segment.Offset, segment.Count, null, null, serializeAsynchronously: true, apm: false)) :
-                base.WriteAsync(buffer, cancellationToken);
-        }
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/IO/Strategies/SyncWindowsFileStreamStrategy.cs:62]: No strong opinion.  If you'd like to consolidate, go for it.  If not, I'm ok keeping the derived types.

**Pull request merged**